### ISSUE TITLE
qwen35 decode dispatch fusion — the lever for MTP parity and tg

### ISSUE THREAD

**Issue by mxaddict** [kryptic-sh/infr#35]
Spun out of #33's perf pass. The MTP spec cycle is now measured end-to-end (`INFR_MTP_TIME=1` + `INFR_PROF2=1`): the m=7 verify forward is **dispatch-count-bound** — 778 dispatches at ~11µs average (lm_head-family 51%, o/down projections 20%, deltanet 15%), costing 23ms vs 9.8ms for a single decode step. Graph build/compile is 0.0ms (pipeline cache), the no-rewind bookkeeping is correct (full-accept cycles run incremental), and reprefill is ~11% of verify time at α=0.95 — so the remaining gap to llama.cpp's own-speedup (theirs 1.68×, ours 1.18-1.21× on the same 7900 XTX) is per-dispatch overhead, not any single kernel.

The fix is **multi-layer/multi-op dispatch fusion** on the qwen35 decode/verify path — collapsing the many small per-layer linears (q/k/v/gate splits, alpha/beta projections, norms) into fewer, wider dispatches. This benefits ALL qwen35 decode (tg128 102 t/s vs llama 163 = 0.62× — same root cause), not just MTP.

Candidates, in rough order:
- Batch the per-layer `ssm_alpha`+`ssm_beta` F32 projections (2 dispatches × 24 layers) into one dispatch per layer or one for all layers (measured est. 1.5-2% alone — only worth it inside a broader fusion pass where the wins compound above bench noise ~3%)
- Fuse norm→projection chains (the AddBias/QkNormRope pattern precedent)
- One indirect multi-layer dispatch for the DeltaNet per-layer recurrence at decode (the batched-MoE routing prologue shows the indirect-dispatch pattern)
- lm_head at 51% of verify GPU time: the m=7 mrow GEMV over 248k vocab — check whether the batched GEMM tier beats mrow at m=7 for out_f this wide

Baselines to beat (4B MTP, 7900 XTX): `tg128 102 t/s | mtp128 120 t/s (1.18-1.21×, α=0.95, draft 31% / verify 57% / catchup 13%)`; llama.cpp same card: tg 163, mtp 273 (1.68×).

Related: the E2B "class-4 dispatch" note and the 2026-07-03 perf-state sweep (remaining gaps class-6/dispatch-bound) — same family of work.

**Comment by digitaloten** [comment]
Candidate 4 (lm_head at m=7) answered on the Metal side — PR #43.

The tied lm_head is Q6_K `out_f = 248320`. At m=2..8 it was taking the multi-row GEMV (`mrv`, tuned on the ≤9216 per-layer projections), where it spawns ~870k simdgroups. Rerouting just the lm_head (gate `out_f >= 65536`) to the `cmm` GEMM tile:

- **qwen3.5-4B-MTP verify: mtp128 1.30× → 1.39× (+8.3%)**, M3 Pro, 3 clean reps
- Qwen3-0.6B self-spec (vocab 151936): +2.8% — generalizes to any wide-`out_f` verify
- Token-identical to greedy and to the old mrv path; parity 94/94; alpha unchanged

So on Metal, the batched GEMM tier does beat the row-tiled GEMV for the lm_head at these m — the mrv register-reuse edge is real only while `out_f` is small.

Note this is the Metal read; the Vulkan `out_f=248320` lm_head at m=7 is worth the same A/B on your side (its GEMV vs coopmat-tile crossover will sit at a different width). The other #35 candidates (ssm_alpha/beta batching, norm→proj fusion, DeltaNet indirect dispatch) are dispatch-COUNT levers — separate from this one, which is pure kernel selection for the single widest op.

**Comment by digitaloten** [comment]
Metal profile note on candidates 1-3 (dispatch fusion) — they are near-futile on Metal, and here is the evidence.

Metal System Trace of qwen3.5-4B decode (M3 Pro), steady state: **GPU is 92% busy** (1954ms busy / 2127ms window; the one 1916ms gap is load/prefill). The 8% idle is entirely the **inter-token CPU gap** — ~2.5ms/token *between* command buffers (one cmdbuf/token), which reconciles exactly with the clean rate (28ms GPU + 2.5ms gap = 30.6ms/token = 32.7 t/s).

Two consequences:
1. The intra-forward dispatch overhead the fusion candidates target is already amortized on Metal by the concurrent replay tape (~0.5µs/dispatch, taped decode). Collapsing the per-layer q/k/v CopyStrided (3/layer) or the ssm_alpha/ssm_beta pair (2/layer) reduces dispatch COUNT, but the GPU is saturated — there is no idle dispatch bubble to reclaim. Estimated Metal win < noise.
2. The only reclaimable time is that 8% inter-token CPU gap, which is NOT dispatch fusion — it is CPU-side per-token work (sampling/detokenize/next-input/commit) serialized against the GPU.

So candidates 1-3 stay valuable on **Vulkan** (your 778-dispatch/11µs verify profile is dispatch-overhead-bound — no concurrent tape there), but on Metal the lever was candidate 4 (#43, work-efficiency) and now that 8% CPU gap. I am profiling the gap next (suspect the 248k-vocab logits readback + host argmax dominates it) and will open anything real separately.

**Comment by digitaloten** [comment]
Profiled the 8% inter-token gap I flagged — it is a dead end on Metal, closing the loop.

`INFR_PROF_DEC` split of the qwen3.5-4B decode step (M3 Pro, steady state), per token:

```
exec (record+submit+gpu+wait)  28.8 ms   (97%)
sample (host argmax over 248k)  0.25 ms
logits download (248k floats)   0.02 ms   <-- essentially free
graph setup (replay tape)       0.13 ms
```

My hypothesis (the 248k-vocab logits readback dominates) was **wrong**: on unified memory the 1 MB "download" is a memcpy from a shared buffer, ~0.02 ms — a device-side argmax would save nothing. Total reclaimable non-GPU work is ~0.4 ms/tok = 1.3%, and the residual gap is inherent command-buffer submit/schedule latency in sequential decode (can not pipeline — token N+1 needs N).

So on Metal the whole decode step is 97% GPU-execute-bound. Candidate 4 (#43) was the lever; candidates 1-3 and the inter-token gap are all below the noise floor here. The dispatch-fusion work remains yours on Vulkan where the concurrent-tape amortization does not apply.
